ns: add compare_weights API with a single model

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92058
* #91639

Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

vkuzo committed:
ns: add compare_weights API with a single model

Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

[ghstack-poisoned]

vkuzo committed:
Update on "ns: add compare_weights API with a single model"


Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

[ghstack-poisoned]

vkuzo committed:
Update on "ns: add compare_weights API with a single model"


Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

[ghstack-poisoned]

vkuzo committed:
Update on "ns: add compare_weights API with a single model"


Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

[ghstack-poisoned]

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -904,6 +904,179 @@ def maybe_remap_node_to_shadow(node):
+    #
+    # Note: this is not super robust, but that's ok because this is
+    # just for legacy customers who depend on the previous two-model version
+    # of this API. New customers should not use this.
```
what does new customers use?

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -904,6 +904,179 @@ def maybe_remap_node_to_shadow(node):
+    #   shadow_wrapper_0_1 = self.shadow_wrapper_0_1(x, w1, b1)
+    #   shadow_0_1 = self.shadow_0_1(shadow_wrapper_0_1, linear)
+    #
+    # algorithm:
```
algorithm is old api and algorithm v2 is the new api? can we just say that in the comment directly?

vkuzo replied:
sorry about that, this is outdated, will remove

jerryzh168 commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -904,6 +904,179 @@ def maybe_remap_node_to_shadow(node):
+        #  to get `_input_scale_1` and `_input_zero_point_1`
+
+        assert len(shadow_n.users) == 1
+        quant_node = list(shadow_n.users.keys())[0]
```
nit: can we look for quant node and then get the weight node from quant_node.args[0], I feel traversing from node to its users is a bit hacky since there might be multiple users etc.

vkuzo replied:
I think that would work for the only case tested here, but it would be more hacky if there are multiple `quantize_per_tensor` functions.  Probably equally hacky to current approach?  I'm flexible, lmk.

vkuzo committed:
Update on "ns: add compare_weights API with a single model"


Summary:

Adds a compare weights NS API using a single model.

Note: this is not intended for wide usage, so testing is limited
to specific functions our customers care about.  The main reason for adding this
is because existing customers of NS are using the old `compare_weights` API,
and we'd like to move everyone to a single-model API style.

Once all the customers are moved over, we can delete all the old NS code.

Test plan:

```
python test/test_quantization.py -k NShadows.test_extract_weights_linear
```

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge

vkuzo commented:
@pytorchbot merge -f "failures are not related"
